src/validate.rs

```diff
@@ -0,0 +1,81 @@
+//! Input validation API (Multi-line editing)
+
+use crate::keymap::Invoke;
+use crate::Result;
+
+/// Input validation result
+#[non_exhaustive]
+pub enum ValidationResult {
+    /// Incomplete input
+    Incomplete,
+    /// Validation fails with an optional error message. User must fix the
+    /// input.
+    Invalid(Option<String>),
```

Comment:
I'm not sure about the validation message. It's fine in your example, but with bigger examples and copy-paste it might be ugly.

Like this (note, I also disallow empty backtick quotes):
```
> SELECT a, b, c, e as ``, f, g, h, i FROM test2 --< unexpected ``
```
Or like this (note, I've pasted the whole chunk of code, so validation is triggered at the end):
```
> SELECT
    a,
    b,
    COUNT(*) as ``
FROM
    test_table
;
 --< unexpected ``
```
Or this:
```
> ELECT
    a,
    b,
    COUNT(*) as ``
FROM
    test_table
; --< Expect: SELECT stmt
```

So to make that helpful we might need to yield position together with validation message and draw validation error in that position, perhaps like this:
```
> SELECT
    a,
    b,
    COUNT(*) as ``
                ^^ unexpected ``
FROM
    test_table
;
```

But I don't think it's worth doing now. The simpler implementation would probably return error span and highlight that span with a color while keeping the validation message at the end. What do you think?

Reply:
Yep. Currently, we cannot do much.
No [bottom bar](https://python-prompt-toolkit.readthedocs.io/en/master/pages/asking_for_input.html#adding-a-bottom-toolbar) !
No [external print](https://github.com/kkawakam/rustyline/pull/235) !
Should we merge ?
Thanks for your reviews.

Reply:
Yes. Let's merge and add more features later.